src/mrpro/operators/LinearOperatorMatrix.py

```diff
@@ -0,0 +1,354 @@
+        ----------
+        x
+            Input tensors. Requires the same number of tensors as the operator has rows.
+
+        Returns
+        -------
+            Output tensors. The same number of tensors as the operator has columns.
+        """
+        return self.H(*x)
+
+    @classmethod
+    def from_diagonal(cls, *operators: LinearOperator):
+        """Create a diagonal LinearOperatorMatrix.
+
+        Create a square LinearOperatorMatrix with the given Linear Operators on the diagonal.
+
+        Parameters
+        ----------
+        operators
+            Sequence of Linear Operators to be placed on the diagonal.
+        """
+        operator_matrix: list[list[LinearOperator]] = [
+            [op if i == j else ZeroOp(False) for j in range(len(operators))] for i, op in enumerate(operators)
+        ]
+        return cls(operator_matrix)
+
+    def operator_norm(
+        self,
+        *initial_value: torch.Tensor,
+        dim: Sequence[int] | None = None,
+        max_iterations: int = 20,
+        relative_tolerance: float = 1e-4,
+        absolute_tolerance: float = 1e-5,
+        callback: Callable[[torch.Tensor], None] | None = None,
+    ) -> torch.Tensor:
+        """Upper bound of operator norm of the Matrix.
+
+        Uses the bounds :math:`||[A, B}^T|||<=sqrt(||A||^2 + ||B||^2)` and :math:`||[A, B]|||<=||A||+||B||`
+        to estimate the operator norm of the matrix by calling operator_norm on each element of the matrix.
+
+        Parameters
+        ----------
+        initial_value
+            Initial value(s) for the power iteration, length should match the number of columns
+            of the operator matrix.
+        dim
+            Dimensions to calculate the operator norm over. Other dimensions are assumed to be
+            batch dimensions. None means all dimensions.
+        max_iterations
+            Maximum number of iterations used in the power iteration.
+        relative_tolerance
+            Relative tolerance for convergence.
+        absolute_tolerance
+            Absolute tolerance for convergence.
+        callback
+            Callback function to be called with the current estimate of the operator norm.
+
+
+        Returns
+        -------
+        Estimated operator norm upper bound.
+        """
+
+        def _singlenorm(op: LinearOperator, initial_value: torch.Tensor):
+            return op.operator_norm(
+                initial_value,
+                dim=dim,
+                max_iterations=max_iterations,
+                relative_tolerance=relative_tolerance,
+                absolute_tolerance=absolute_tolerance,
+                callback=callback,
+            )
+
+        if len(initial_value) != self.shape[1]:
+            raise ValueError('Initial value should have the same length as the operator has columns.')
+        norms = torch.tensor(
+            [[_singlenorm(op, iv) for op, iv in zip(row, initial_value, strict=True)] for row in self._operators]
```

Comment:
maybe I misunderstand the code, but for example, for this operator,
```
Op = [A,0,0]
     [D,B,0]
     [0,0,C]
```
having the estimate given by ||Op|| < sqrt(||A||**2 + .... ||D||**2) seems to be quite bad, see the following example, where you can change `lower_diagonal_flag  ` to either include a lower-diagonal operator or not. Regardless of that, the estimate seems rather bad, shower. For example, running

```
import torch 
from math import sqrt

torch.manual_seed(0)
n=32

lower_diagonal_flag = False     
A = torch.randn(n,n)
B = torch.randn(n,n)
C = torch.randn(n,n)

#block-diagonal matrix
operator_list = [A,B,C]
Mat = torch.block_diag(*operator_list)

operator_names = ["A","B","C"]

if lower_diagonal_flag:
    D = torch.randn(n,n)
    operator_list.append(D)
    operator_names.append("D")
    
    #put some matrix in the lower diagonal
    Mat[n:n+n,:n] = D 


Mat_norm = torch.linalg.matrix_norm(Mat,ord=2)
for op_name, operator in zip(operator_names, operator_list):
    print("op norm of ",op_name, torch.linalg.matrix_norm(operator,ord=2).item())

def upper_bound_op_norm(operators):
    return sqrt(sum([torch.linalg.matrix_norm(op,ord=2).item()**2 for op in operators]))

print("op norm of Mat", Mat_norm.item())
print("est upper bound: ",upper_bound_op_norm(operator_list))
```
gives me

```
op norm of  A 11.429673194885254
op norm of  B 10.949110984802246
op norm of  C 10.883465766906738
op norm of Mat 11.42967414855957
est upper bound:  19.20859931902445
```
which also shows that, if we have purely block-diagonal operator, taking the max of the operator-norms of the operators gives the exact operator norm (we had also a test-example for that).

Thus, maybe it is worth distinguishing for the cases where we have a block-diagonal operator or not?

Reply:
additionally, if we still opt for calculating only an upper bound for the operator norm instead of the actual operator norm, we could/should at least raise a warning?

Reply:
you are right, for horzontally stacked operators, it could be the max.
I just thought the result has to be bigger than either, so sum .. but using the max would of course be better to achieve that.
so root-sum-square of rows and max of cols.
this would make the bound a bit tighter, especially for diagonl operators.

```python
import einops
norms=torch.linalg.matrix_norm(einops.rearrange(Mat,"(r n) (c m)-> r c n m", n=n, m=n),ord=2)
print('norms',norms)
norm=norms.square().sum(-2).sqrt().amax(-1)
print('tighter', norm.item())
```
```
op norm of Mat 14.396413803100586
old upper bound:  22.011787375336397
norms       [[11.4297,  0.0000,  0.0000],
            [10.7493, 10.9491,  0.0000],
            [ 0.0000,  0.0000, 10.8835]]
tighter bound 15.690311431884766
```

Reply:
good catch!